cpp/src/transform/transform.cpp

```diff
@@ -166,91 +167,285 @@ jitify2::Kernel get_kernel(mutable_column_view output,
+                                            column_type_names(output_columns),
+                                            reflect_input_columns(base_column_size, input_columns),
+                                            num_user_data)),
+                    cuda_source)
+    ->configure_1d_max_occupancy(0, 0, nullptr, stream.value());
+}
+
+jitify2::ConfiguredKernel build_span_kernel(std::string const& kernel_name,
+                                            size_type base_column_size,
+                                            std::vector<std::string> const& span_outputs,
+                                            std::vector<column_view> const& input_columns,
+                                            int32_t num_user_data,
+                                            std::string const& udf,
+                                            bool is_ptx,
+                                            rmm::cuda_stream_view stream,
+                                            rmm::device_async_resource_ref mr)
+{
+  auto const cuda_source =
     is_ptx ? cudf::jit::parse_single_function_ptx(
-               udf, "GENERIC_TRANSFORM_OP", build_ptx_params(output, inputs))
+               udf,
+               "GENERIC_TRANSFORM_OP",
+               build_ptx_params(span_outputs, column_type_names(input_columns), num_user_data))
            : cudf::jit::parse_single_function_cuda(udf, "GENERIC_TRANSFORM_OP");
 
-  auto kernel = get_kernel(output, inputs, cuda_source)
-                  ->configure_1d_max_occupancy(0, 0, nullptr, stream.value());
+  return get_kernel(jitify2::reflection::Template(kernel_name)
+                      .instantiate(
+                        build_jit_typenames(span_outputs,
+                                            {},
+                                            reflect_input_columns(base_column_size, input_columns),
+                                            num_user_data)),
+                    cuda_source)
+    ->configure_1d_max_occupancy(0, 0, nullptr, stream.value());
+}
+
+void launch_column_output_kernel(jitify2::ConfiguredKernel& kernel,
+                                 std::vector<mutable_column_view> const& output_columns,
+                                 std::vector<column_view> const& input_columns,
+                                 std::optional<void*> user_data,
+                                 rmm::cuda_stream_view stream,
+                                 rmm::device_async_resource_ref mr)
+{
+  auto [output_handles, outputs] =
+    column_views_to_device<mutable_column_device_view, mutable_column_view>(
+      output_columns, stream, mr);
+  auto [input_handles, inputs] =
+    column_views_to_device<column_device_view, column_view>(input_columns, stream, mr);
+  auto user_data_host =
+    user_data.has_value() ? std::vector<void*>{user_data.value()} : std::vector<void*>{};
+  auto device_user_data = to_device_vector(user_data_host, stream, mr);
```

Comment:
Still wishing this was not a vector.

Reply:
the kernel takes arrays

Reply:
Does it have to?

Reply:
yes, the accessors (`user_data_accessor` in this case) use indices into the arrays to unpack their data

Reply:
also, the vector is used to upload the pointer to device, I'd rather not have to create a `unique_ptr` for device memory.

Reply:
I guess I'm not understanding why we cannot just pass the pointer like any other pass-by-value primitive type. The pointer should not be wrapped with a unique_ptr since we don't own the memory and should never free it. Its just an 8-byte value we are passing through from the transform call all the way to the kernel.

Reply:
The accessors are indexed, so they need to unpack their data from device arrays. 
And by `unique_ptr` I mean one to temporarily store the user_data pointer value to be passed to the kernel as a pointer.
I'll make the `user_data_accessor` a special-case and make it unindexed.